### PR TITLE
bugfix/polling-multiple-redraws

### LINKED ISSUE #20483: Dashboards: Animation lost on HighchartsComponent on polling
#### Expected behaviour
Charts should animate smoothly like [they do with polling in the data module](https://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/data/livedata-rows/). Only one chart redraw should be performed.

#### Actual behaviour
Animation is lost. On each iteration, the chart is redrawn 4 times, the first one without data points, causing animation to be lost, and also a waste of CPU due to excess redrawing.

#### Live demo with steps to reproduce
https://jsfiddle.net/highcharts/vu3fny4h/
Look at the console. When we start polling, we immediately have two redraws. Those are not the focus of this issue.

Then when the first polling cycle runs, we have four calls to `redraw`.

#### Product version
v1.2.1

#### Debugging results
It looks like the DataTable is either too eager to spark events, or the polling mechanism should not pick them all up.
* For each polling cycle, `DataTable.deleteRows` runs, emitting an event that ultimately redraws all connected components with empty data. That should not happen, because we know that new data is coming in.
* `DataTable.setModifier` also sparks events that redraw the components through the `afterSetModifier` event.

When I comment out these two event emitters, the chart only redraws once, and the animation works as intended:
```diff
diff --git a/ts/Data/DataTable.ts b/ts/Data/DataTable.ts
index 704bd38b55..b9d7522b15 100644
--- a/ts/Data/DataTable.ts
+++ b/ts/Data/DataTable.ts
@@ -502,6 +502,7 @@ class DataTable implements DataEvent.Emitter {
             );
         }
 
+        /*
         table.emit({
             type: 'afterDeleteRows',
             detail: eventDetail,
@@ -509,6 +510,7 @@ class DataTable implements DataEvent.Emitter {
             rowIndex: (rowIndex || 0),
             rows: deletedRows
         });
+        */
 
         return deletedRows;
     }
@@ -1459,12 +1461,14 @@ class DataTable implements DataEvent.Emitter {
 
         return promise
             .then((table): this => {
+                /*
                 table.emit({
                     type: 'afterSetModifier',
                     detail: eventDetail,
                     modifier,
                     modified: table.modified
                 });
+                */
                 return table;
             })['catch']((error): this => {
                 table.emit({
```

In summary, we should economize with the events causing chart redraws. Either we shouldn't emit events in some cases, or the handlers should be adjusted to cancel them. Furthermore, this probably applies to all components, and for those that don't have the animation issue, it will still be a CPU sink.

### PR DESCRIPTION
**Pull request by pawellysy**
Fixed #20483, there were multiple redraws of the chart in `HighchartsComponent` when polling.
FYI @sebastianbochan

### CODE REVIEW

**Review by TorsteinHonsi — APPROVED** [review]
Thanks, looks good to me 👍

File: ts/Data/Connectors/CSVConnector.ts [hunk 1]
```diff
@@ -149,6 +147,8 @@ class CSVConnector extends DataConnector {
             )
             .then((csv): Promise<string> => {
                 if (csv) {
+                    // If already loaded, clear the current rows
+                    table.deleteColumns();
```
Comment (karolkolodziej) [inline ts/Data/Connectors/CSVConnector.ts:151]: Don't we need the same in Google and HTML connectors?

**Review by karolkolodziej — APPROVED** [review]

File: ts/Data/Connectors/JSONConnector.ts [hunk 2]
```diff
@@ -139,14 +137,15 @@ class JSONConnector extends DataConnector {
...
+                    // If already loaded, clear the current rows
+                    table.deleteColumns();
                     converter.parse({ data });
                     table.setColumns(converter.getTable().getColumns());
                 }
-                return connector
-                    .setModifierOptions(dataModifier)
-                    .then((): Array<Array<number|string>> => data);
+                await connector.setModifierOptions(dataModifier);
+                return data;
```
Comment (bre1470) [inline ts/Data/Connectors/JSONConnector.ts:148]: No async/await in the data layer as it does not work with the polyfill for old browsers.

**Review by bre1470 — CHANGES_REQUESTED** [review]
It would have been better, if the indent would been kept the same. I do not see any difference. The Promise pipe needs a fix to be backward compatible though.

**Comment by pawellysy** [comment]
@bre1470 I reversed the `if` statement condition and added the early return, which I like more, than the nested indentation. 
If you prefer keeping it the way it was, I can revert the change.

File: ts/Data/Connectors/JSONConnector.ts (hunk 2 above)
Comment (TorsteinHonsi) [reply ts/Data/Connectors/JSONConnector.ts]: @bre1470 Is it the es5 builds that don't work? It would be nice to be able to do async/await.

Comment (bre1470) [reply ts/Data/Connectors/JSONConnector.ts]: Exactly, ES5 will have problems as this is a language feature and not API.

**Comment by bre1470** [comment]
@pawellysy It is fine, just difficult to notice changes.

**Review by bre1470 — APPROVED** [review]
Thanks! Looks good. 👍

**Pull request merged**